File: tests/algorithms/x/Pascal/maths/monte_carlo_dice.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
type RealArray = array of real;
type IntArray = array of integer;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
function list_real_to_str(xs: array of real): string;
var i: integer;
begin
  Result := '[';
  for i := 0 to High(xs) do begin
    Result := Result + FloatToStr(xs[i]);
    if i < High(xs) then Result := Result + ' ';
  end;
  Result := Result + ']';
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  lcg_seed: integer;
  num_dice: integer;
  num_throws: integer;
  x: real;
function lcg_rand(): integer; forward;
function roll(): integer; forward;
function round2(x: real): real; forward;
function throw_dice(num_throws: integer; num_dice: integer): RealArray; forward;
procedure main(); forward;
function lcg_rand(): integer;
begin
  lcg_seed := ((lcg_seed * 1103515245) + 12345) mod 2147483648;
  exit(lcg_seed);
end;
function roll(): integer;
var
  roll_rv: real;
  roll_r: real;
begin
  roll_rv := Double(lcg_rand());
  roll_r := (roll_rv * 6) / 2.147483648e+09;
  exit(1 + Trunc(roll_r));
end;
function round2(x: real): real;
var
  round2_y: real;
  round2_z: integer;
begin
  round2_y := (x * 100) + 0.5;
  round2_z := Trunc(round2_y);
  exit(Double(round2_z) / 100);
end;
function throw_dice(num_throws: integer; num_dice: integer): RealArray;
var
  throw_dice_count_of_sum: array of integer;
  throw_dice_max_sum: integer;
  throw_dice_i: integer;
  throw_dice_t: integer;
  throw_dice_s: integer;
  throw_dice_d: integer;
  throw_dice_probability: array of real;
  throw_dice_p: real;
begin
  throw_dice_count_of_sum := [];
  throw_dice_max_sum := (num_dice * 6) + 1;
  throw_dice_i := 0;
  while throw_dice_i < throw_dice_max_sum do begin
  throw_dice_count_of_sum := concat(throw_dice_count_of_sum, IntArray([0]));
  throw_dice_i := throw_dice_i + 1;
end;
  throw_dice_t := 0;
  while throw_dice_t < num_throws do begin
  throw_dice_s := 0;
  throw_dice_d := 0;
  while throw_dice_d < num_dice do begin
  throw_dice_s := throw_dice_s + roll();
  throw_dice_d := throw_dice_d + 1;
end;
  throw_dice_count_of_sum[throw_dice_s] := throw_dice_count_of_sum[throw_dice_s] + 1;
  throw_dice_t := throw_dice_t + 1;
end;
  throw_dice_probability := [];
  throw_dice_i := num_dice;
  while throw_dice_i < throw_dice_max_sum do begin
  throw_dice_p := (Double(throw_dice_count_of_sum[throw_dice_i]) * 100) / Double(num_throws);
  throw_dice_probability := concat(throw_dice_probability, [round2(throw_dice_p)]);
  throw_dice_i := throw_dice_i + 1;
end;
  exit(throw_dice_probability);
end;
procedure main();
var
  main_result_: RealArray;
begin
  lcg_seed := 1;
  main_result_ := throw_dice(10000, 2);
  writeln(list_real_to_str(main_result_));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  lcg_seed := 1;
  main();
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/maths/monte_carlo_dice.clj
(ns main (:refer-clojure :exclude [lcg_rand roll round2 throw_dice main]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare lcg_rand roll round2 throw_dice main)

(def ^:dynamic main_result nil)

(def ^:dynamic roll_r nil)

(def ^:dynamic roll_rv nil)

(def ^:dynamic round2_y nil)

(def ^:dynamic round2_z nil)

(def ^:dynamic throw_dice_count_of_sum nil)

(def ^:dynamic throw_dice_d nil)

(def ^:dynamic throw_dice_i nil)

(def ^:dynamic throw_dice_max_sum nil)

(def ^:dynamic throw_dice_p nil)

(def ^:dynamic throw_dice_probability nil)

(def ^:dynamic throw_dice_s nil)

(def ^:dynamic throw_dice_t nil)

(def ^:dynamic main_lcg_seed 1)

(defn lcg_rand []
  (try (do (alter-var-root (var main_lcg_seed) (fn [_] (mod (+ (* main_lcg_seed 1103515245) 12345) 2147483648))) (throw (ex-info "return" {:v main_lcg_seed}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn roll []
  (binding [roll_r nil roll_rv nil] (try (do (set! roll_rv (double (lcg_rand))) (set! roll_r (/ (* roll_rv 6.0) 2147483648.0)) (throw (ex-info "return" {:v (+ 1 (long roll_r))}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn round2 [round2_x]
  (binding [round2_y nil round2_z nil] (try (do (set! round2_y (+ (* round2_x 100.0) 0.5)) (set! round2_z (long round2_y)) (throw (ex-info "return" {:v (/ (double round2_z) 100.0)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn throw_dice [throw_dice_num_throws throw_dice_num_dice]
  (binding [throw_dice_count_of_sum nil throw_dice_d nil throw_dice_i nil throw_dice_max_sum nil throw_dice_p nil throw_dice_probability nil throw_dice_s nil throw_dice_t nil] (try (do (set! throw_dice_count_of_sum []) (set! throw_dice_max_sum (+ (* throw_dice_num_dice 6) 1)) (set! throw_dice_i 0) (while (< throw_dice_i throw_dice_max_sum) (do (set! throw_dice_count_of_sum (conj throw_dice_count_of_sum 0)) (set! throw_dice_i (+ throw_dice_i 1)))) (set! throw_dice_t 0) (while (< throw_dice_t throw_dice_num_throws) (do (set! throw_dice_s 0) (set! throw_dice_d 0) (while (< throw_dice_d throw_dice_num_dice) (do (set! throw_dice_s (+ throw_dice_s (roll))) (set! throw_dice_d (+ throw_dice_d 1)))) (set! throw_dice_count_of_sum (assoc throw_dice_count_of_sum throw_dice_s (+ (nth throw_dice_count_of_sum throw_dice_s) 1))) (set! throw_dice_t (+ throw_dice_t 1)))) (set! throw_dice_probability []) (set! throw_dice_i throw_dice_num_dice) (while (< throw_dice_i throw_dice_max_sum) (do (set! throw_dice_p (/ (* (double (nth throw_dice_count_of_sum throw_dice_i)) 100.0) (double throw_dice_num_throws))) (set! throw_dice_probability (conj throw_dice_probability (round2 throw_dice_p))) (set! throw_dice_i (+ throw_dice_i 1)))) (throw (ex-info "return" {:v throw_dice_probability}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn main []
  (binding [main_result nil] (do (alter-var-root (var main_lcg_seed) (fn [_] 1)) (set! main_result (throw_dice 10000 2)) (println (str main_result)))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
